src/plugins/csp_handler/server/csp_handlers.ts

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { ConfigurationClient } from '../../application_config/server';
+import {
+  CoreSetup,
+  IScopedClusterClient,
+  Logger,
+  OnPreResponseHandler,
+  OnPreResponseInfo,
+  OnPreResponseToolkit,
+  OpenSearchDashboardsRequest,
+} from '../../../core/server';
+
+const CSP_RULES_CONFIG_KEY = 'csp.rules';
+
+/**
+ * This function creates a pre-response handler to dynamically set the CSP rules.
+ * It give precedence to the rules from application config plugin over those from YML.
+ * In case no value from application config, it will ensure a default frame-ancestors is set.
+ *
+ * @param core Context passed to the plugins `setup` method
+ * @param cspHeader The CSP header from YML
+ * @param getConfigurationClient The function provided by application config plugin to retrieve configurations
+ * @param logger The logger
+ * @returns The pre-response handler
+ */
+export function createCspRulesPreResponseHandler(
+  core: CoreSetup,
+  cspHeader: string,
+  getConfigurationClient: (scopedClusterClient: IScopedClusterClient) => ConfigurationClient,
```

Comment:
not blocker for this PR, but we expose getConfigurationClient to plugins, right? If yes, then why do they need pass `IScopedClusterClient` to get `ConfigurationClient`. After all plugin or component who implements interfaces, will build their own client.

Reply:
Good call out! Let me elaborate more.

When an external configuration client is registered, there is no need to use `IScopedClusterClient`. We just return the external client. 
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/application_config/server/plugin.ts#L54

The reason why `IScopedClusterClient` is needed for the default OpenSearch backend is that we need to check permissions by using the `asCurrentUser` from the scoped client. This is only available from request context and this is also the reason why we need a function instead of the client. https://github.com/opensearch-project/OpenSearch-Dashboards/pull/5641/files#r1517079679

https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/application_config/server/opensearch_config_client.ts#L50

Reply:
I understand that. suppose I am implementing my own Configuration Client in external component say Plugin. 
```
Myconfig: ConfigurationClient {
...
}
```

in plugin setup register the configuration client 
```
applicationconfig.registerConfigurationClient(Myconfig);
```

Now I'm trying to get Configuraton Client using `getConfigurationClient`.
How to call this API without IScopedClusterClient as IScopedClusterClient is required param of getConfigurationClient. How the plugin would get IScopedClusterClient? And they don't need to pass IScopedClusterClient, right? Because when they implement their own ConfiguratonClient, IScopedClusterClient doesn't even come into picture. Correct me if I'm wrong or missing something here?

Reply:
Customers could pass null to this field when there is no such thing. The if check will be true since there will be an external client registered.

https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/application_config/server/plugin.ts#L54

Reply:
instead forcing user to pass null, I think it would be better to make optional parameter? anyway this is not in scope of this PR. This is something we might fix in ApplicationConfig. If you agree to make optional parameter, you can create an issue and resolve this conversation.

Reply:
Great suggestion!

Created the follow up issue https://github.com/opensearch-project/OpenSearch-Dashboards/issues/6100

@bandinib-amzn